Quantum-Control-Applications/ESR/cpmg.py

```diff
@@ -0,0 +1,213 @@
+from qm import SimulationConfig
+from qm.qua import *
+from qm import LoopbackInterface
+from qm.QuantumMachinesManager import QuantumMachinesManager
+from configuration import *
+import matplotlib.pyplot as plt
+from qm.simulate.credentials import create_credentials
+
+###################
+# The QUA program #
+###################
+
+tau_min = 2000 // 4  # tau is half the time between pi pulses
+tau_max = 5e4 // 4  # tau is half the time between pi pulses
+tau_array = np.round(np.exp(np.linspace(np.log(tau_min), np.log(tau_max), 30)))  # outputs results in float
+
+tau_array_int = tau_array.astype(int).tolist()
+
+num_taus = len(tau_array)
+
+latest_echo = 50e6 // 4  # this is the maximum amount of time I will run CPMG
+
+num_echo_array = (np.floor(latest_echo / (2 * tau_array))).astype(int).tolist()
+
+total_num_echoes = np.sum(num_echo_array)  # total number of echoes played in the whole sequence
+
+n_avg = 100
+
+with program() as cpmg:
+
+    n = declare(int)
+    n_st = declare_stream()
+
+    I = declare(fixed)
+    Q = declare(fixed)
+
+    I_st = declare_stream()
+    Q_st = declare_stream()
+
+    tau_st = declare_stream()
+    i_echo_st = declare_stream()
+
+    tau_array_q = declare(int, value=tau_array_int)
+    tau = declare(int)
+    i_tau = declare(int)
+
+    num_echo_array_q = declare(int, value=num_echo_array)
+    i_echo = declare(int)
```

Comment:
Is this variable the same as 'iteration' that you used in the SC codes ? If so, I would keep iteration which seems clearer

Reply:
it is not iteration, I have `n` for iteration. In this case i_echo is the variable that I loop over to create a CPMG that has a variable number of pi pulses in the sequence. We called it i_echo because is related to the number of echos you will collect in a single run

Reply:
Yes correct, my bad...